File: 25-State/src/cpp/main.cpp
#include <iostream>
#include "State.h"
#include "GumballMachine.h"
#include "NoQuarterState.h"
#include "HasQuarterState.h"
#include "SoldState.h"
#include "SoldOutState.h"

// Defined here, where every concrete state class is complete, which breaks
// the GumballMachine <-> concrete-state definition cycle.
GumballMachine::GumballMachine(int count) : count_(count) {
    no_quarter_ = std::make_unique<NoQuarterState>(*this);
    has_quarter_ = std::make_unique<HasQuarterState>(*this);
    sold_ = std::make_unique<SoldState>(*this);
    sold_out_ = std::make_unique<SoldOutState>(*this);
    current_ = count > 0 ? no_quarter_.get() : sold_out_.get();
}

int main() {
    std::cout << "=== State Pattern Demo (C++) ===\n\n";

    GumballMachine machine(2);
    std::cout << machine.status() << "\n";

    std::cout << "\nBuy first gumball:\n";
    machine.insert_quarter();
    machine.turn_crank();
    std::cout << machine.status() << "\n";

    std::cout << "\nBuy second gumball (empties the machine):\n";
    machine.insert_quarter();
    machine.turn_crank();
    std::cout << machine.status() << "\n";

    std::cout << "\nTry to buy when sold out:\n";
    machine.insert_quarter();
    machine.turn_crank();

    std::cout << "\nSame actions behaved differently in each state \u2014\n";
    std::cout << "no conditionals, just the current State object.\n";

    return 0;
}

File: 25-State/src/rust/main.rs
mod state;
mod no_quarter_state;
mod has_quarter_state;
mod sold_state;
mod sold_out_state;
mod gumball_machine;

use gumball_machine::GumballMachine;

fn main() {
    println!("=== State Pattern Demo (Rust) ===\n");

    let mut machine = GumballMachine::new(2);
    println!("{}", machine.status());

    println!("\nBuy first gumball:");
    machine.insert_quarter();
    machine.turn_crank();
    println!("{}", machine.status());

    println!("\nBuy second gumball (empties the machine):");
    machine.insert_quarter();
    machine.turn_crank();
    println!("{}", machine.status());

    println!("\nTry to buy when sold out:");
    machine.insert_quarter();
    machine.turn_crank();

    println!("\nSame actions behaved differently in each state —");
    println!("no conditionals, just the current State object.");
}
